### DISCUSSION TITLE
Return button keeps change

### DISCUSSION THREAD

**Discussion by olliw42** [EdgeTX/edgetx#1162] [Ideas]
observed on a Jumper T16:

when doing changes to settings and one is returning by pressing the RTN button, then the changes are accepted nevertheless! They should be accepted only upon a press of return.

This is pretty annoying since it is essentially impossible to jump out from wrong changes.

**Comment by raphaelcoeffic** [comment]
I think this is a generic issue on almost all fields: there is no "cancel". I'd have to look into it, but it seems it's going to be a pain to add (lots of things to check). As far as I've seen, the fields are edited directly (changed directly in the model / radio settings). It would require modding libopenui to support "copy-on-edit", and "write-back-on-accept", which it does not at all right now.

Let's tackle this a bit later. But you're right, this is often annoying.

**Comment by pfeerick** [comment]
I raised this with OpenTX devs, as this is, and has always been standard behaviour there, and they said... no. I agree that it is often annoying, as you also can't change things live, and then hit return to revert... but I also get it is a bugger of a feature to implement across all the different TX types and interfaces.

**Comment by olliw42** [comment]
~~it is NOT the behavior of otx2,3 !~~ wasn't a correct statement

I frankly would not agree with postponing this, IMHO it is super annoying and just soo standard for any GUI type of thing that it is IMHO just clear what the responses of all users would be concerning if etx2.4 is worth it ...

IMHO this should be a blocker for any release

my 2 cents

**Comment by raphaelcoeffic** [comment]
@olliw42 ok, I’ll re-check.

**Comment by pfeerick** [comment]
I suggest you revisit OTX 2.3, as you will find it *is* the behaviour, and has always been that behavior of OTX since I started using it in the 1.x days with my Turnigy 9X!! It's not new! If you make any changes to a setting whatsover, pressing return will save it, the same as pressing enter will. Furthermore, you can see the devs say *why* they refused to change from the current behaviour, to what you are suggesting, here. https://github.com/opentx/opentx/issues/8369

**Comment by olliw42** [comment]
> I suggest you revisit OTX 2.3, as you will find it _is_ the behaviour

that's indeed correct, my fault 

but it doesn't change my opinion how it should be ... 

the arguments provided by some devs in the ref you gave, like "that's how every R/C radio out there works" or "it would cause massive confusion with users especially since everything IT is moving to a similar "instant save" model nowadays", are - IMHO - just nonsense and do not appear to be generally true

**Comment by pfeerick** [comment]
Indeed... that's why I gave up on the issue... 

To be fair, I can see both arguments being true... as it would mean a change in a behavior that those of us who have been using it for for a long time could lead to confusion or even a plane crashing... because as it was pointed out, sometimes it is necessary to make those changes in flight for tuning, and having return now suddenly revert a setting could be quite disastrous since until now it was automatic/instinctive that that all that would happen is the field being edited left, not the edit reverted. 

Which is why I then suggested it should be an option, but the question naturally then was 'who would implement the feature' ... I would have preferred they get 2.4 out the door before working on 'would like' features. Given the amount of work required, and that it would almost certainly be a blocker for re-integration into opentx post 2.4, I do not think this should be considered or implemented yet, and there are much more important issues to be worked on given the proposed release date is just over a month away. 

Unless I am mistaken the main goal of this first release is to basically do what was promised - OpenTX 2.4 with colour touch screen support. After that, enhancements like this can/should then be assessed, discussed, and assigned to milestones to clearly indicate the direction things are going.

**Comment by JimB40** [comment]
To be closed imho. Or moved to dicuss.

**Comment by pfeerick** [comment]
Closed, no, not a chance - this is why partly why it was raised here, as that is what happened over at OTX. Moved, most likely, as will quite a few issues I think ;)  Correction - not closed, it's just stagnant.

**Comment by MervDale** [comment]
I'll bite , doesn't having to use the Enter key negate somewhat the functionality of a Touchscreen,  one example being  the Enable Box, the mere fact that touching this box selects it but if you then are required to press the Enter Key you will deselect it,?

**Comment by pfeerick** [comment]
Please don't... there's only like three transmitters out of 20 that have touch screen... so that basically throws that argument out the window :-P You might want to read up on this issue a bit - it's in relation to the historical behaviour of RTN, not ENT, on saving of data.

**Comment by JimB40** [comment]
Besides it's like debating what's better Christmas or Easter ... :)
I'd really leave it as it is ATM
Instant change (no need to confirm value change) is done across all OS with some excptions

There are just some bugs to fix to make it consistent
Example is Slider:
- with keys: when you scroll to next value and press EXIT -> value is changed :)

**Reply by pfeerick** [reply]
And don't forget choice lists (i.e. your mixes/input source list)... if you press RTN instead of ENT it will revert to the prior value. That was a deliberate change though, which we need to consider the ramifications of.

**Comment by MervDale** [comment]
This issue as I see it started before touch screens and ETX in that there is a difference as to how this works depending on whether you have a touch screen radio or not!

With the touch screen radio you can select the part to be edited by touch and in doing so automatically puts it into edit mode.  This is not the case with a non touch screen, you also use the enter button to enter edit mode.

The touch screen by design is supposed to make things quicker and easier and the way ETX is now I don't even use the return button to get out the edit mode,  I just touch the top left corner of the screen knowing that what I edited is automatically saved.  This is also the same for my Horus X10S the only difference it is a slower clumsier process. 
Maybe an auto backup mode is a good idea every time you go into edit mode but is it required. 
 Maybe we need HAL " ARE YOU SURE YOU WANT TO DO THIS DAVE" 

Everyone that has been using OTX and now ETX understands the basics by now and understand it is not the same as using a Jeti or a Spektrum radio.

My 2 cents worth  ;-Q

**Reply by pfeerick** [reply]
I think what has happened now that we are using touch on touch-capable radios we assume that when we exit all changes are preserved, whereas what olliew42 was suggesting is that using RTN on a page would let you exit without saving changes. Now, yes, there could indeed be a "do you want to save these changes" type deal... but the underlying issue was then "how do we get to the previous settings since it's edited in place". 

As an user of OTX, and now ETX, I personally feel things are right as behaviour - RTN is basically leaving the screen, not undo. In fact, on a radio that I know of that does follow this sort of mechanism - OrangeRX TX6i - it is downright confusing, and leads to settings being lost all the time - as keep forgetting it needs a long press of ENT for the live settings to be saved. None of my other transmitters - be it spektrum, Walkera, OTX, etc,  use this long press or ENT does one thing, RTN does another model. However, I am only one of many. ;)

**Comment by MervDale** [comment]
There is an old adage that I like.   " If it ain't broke don't fix it ! "

**Reply by pfeerick** [reply]
Who decides it's not broken? :grin:

**Comment by MervDale** [comment]
Certainly not the minority,  I have been down that road many times and having to fix a system that was running well until some one decided to do a Frank Sinatra.  ;-Q
